### PR TITLE
More metrics for on-disk temporary data

### LINKED ISSUE #40206: More metrics for on-disk temporary data
**Use case**

Track the amount of space used for temporary data.

**Describe the solution you'd like**

We already have the following ProfileEvents:

`ExternalSortWritePart`
`ExternalSortMerge`
`ExternalAggregationWritePart`
`ExternalAggregationMerge`
`ExternalAggregationCompressedBytes`
`ExternalAggregationUncompressedBytes`

But there is no information about CompressedBytes and UncompressedBytes for external sort.
And there is no information about external JOIN at all.

Let's add it.
Let's also add ProfileEvents for the values (num files, compressed bytes, uncompressed bytes) of all operations in total.

Let's also add `CurrentMetrics` for the number of currently used files, compressed bytes, and uncompressed bytes, for all of the three types of operations, and for all the operations in total.

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Add more metrics for on-disk temporary data, close https://github.com/ClickHouse/ClickHouse/issues/40206

TODO:
- [x] Profile Events for `ExternalSortCompressedBytes`/`ExternalSortUncompressedBytes`
- [x] Profile Events for partial_merge_join (for full_sorting_join events for sort are used)
- [x] ProfileEvents for total sort/agg/join
- [x] CurrentMetrics

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Formats/TemporaryFileStream.cpp [hunk 1]
```diff
@@ -39,6 +39,10 @@ void TemporaryFileStream::write(const std::string & path, const Block & header,
         output.write(block);
 
     compressed_buf.finalize();
+    output.flush();
+    file_buf.next();
+
+    return Stat{compressed_buf.count(), file_buf.count()};
```
Comment (antonio2368) [inline src/Formats/TemporaryFileStream.cpp:45]: Why not use `CompressedWriteBuffer` methods `getCompressedBytes` and `getUncompressedBytes`?
This really confused me, wasn't sure if you switched them around but it turned out correct.

Comment (vdimir) [reply src/Formats/TemporaryFileStream.cpp]: Good idea, updated BufferingToFileTransform and TemporaryFileStream to use this methods

File: src/Common/TemporaryFile.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+TemporaryFileOnDisk::~TemporaryFileOnDisk()
+{
+    try
+    {
+        if (disk && !filepath.empty() && disk->exists(filepath))
+            disk->removeRecursive(filepath);
+    }
+    catch (...)
+    {
```
Comment (antonio2368) [inline src/Common/TemporaryFile.cpp:45]: Maybe log something or would it be too much for all files?

File: src/Interpreters/Aggregator.cpp [hunk 3]
```diff
@@ -1469,40 +1477,26 @@ bool Aggregator::executeOnBlock(Columns columns,
...
+void Aggregator::writeToTemporaryFile(AggregatedDataVariants & data_variants, size_t max_temp_file_size) const
 {
     Stopwatch watch;
     size_t rows = data_variants.size();
 
-    auto file = createTemporaryFile(tmp_path);
-    const std::string & path = file->path();
+    auto file = createTempFile(max_temp_file_size);
+
+    std::string path = file->getPath();
```
Comment (antonio2368) [inline src/Interpreters/Aggregator.cpp:1494]:
```suggestion
    const auto & path = file->getPath();
```

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**